shell/platform/darwin/ios/framework/Source/FlutterMetalLayer.mm

```diff
@@ -0,0 +1,402 @@
+  self->_presented = YES;
+}
+
+- (void)dealloc {
+  if (!_presented) {
+    [_layer returnTexture:self->_texture];
+  }
+}
+
+- (void)addPresentedHandler:(nonnull MTLDrawablePresentedHandler)block {
+  FML_LOG(WARNING) << "FlutterMetalLayer drawable does not implement addPresentedHandler:";
+}
+
+- (void)presentAtTime:(CFTimeInterval)presentationTime {
+  FML_LOG(WARNING) << "FlutterMetalLayer drawable does not implement presentAtTime:";
+}
+
+- (void)presentAfterMinimumDuration:(CFTimeInterval)duration {
+  FML_LOG(WARNING) << "FlutterMetalLayer drawable does not implement presentAfterMinimumDuration:";
+}
+
+@end
+
+@implementation FlutterMetalLayer
+
+@synthesize preferredDevice = _preferredDevice;
+@synthesize device = _device;
+@synthesize pixelFormat = _pixelFormat;
+@synthesize framebufferOnly = _framebufferOnly;
+@synthesize colorspace = _colorspace;
+@synthesize wantsExtendedDynamicRangeContent = _wantsExtendedDynamicRangeContent;
+
+- (instancetype)init {
+  if (self = [super init]) {
+    _preferredDevice = MTLCreateSystemDefaultDevice();
+    self.device = self.preferredDevice;
+    self.pixelFormat = MTLPixelFormatBGRA8Unorm;
+    _availableTextures = [[NSMutableSet alloc] init];
+
+    _displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(onDisplayLink:)];
+    [self setMaxRefreshRate:[DisplayLinkManager displayRefreshRate] forceMax:NO];
+    [_displayLink addToRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(didEnterBackground:)
+                                                 name:UIApplicationDidEnterBackgroundNotification
+                                               object:nil];
+  }
+  return self;
+}
+
+- (void)dealloc {
+  [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)setMaxRefreshRate:(double)refreshRate forceMax:(BOOL)forceMax {
+  // This is copied from vsync_waiter_ios.mm. The vsync waiter has display link scheduled on UI
+  // thread which does not trigger actual core animation frame. As a workaround FlutterMetalLayer
+  // has it's own displaylink scheduled on main thread, which is used to trigger core animation
+  // frame allowing for 120hz updates.
+  if (!DisplayLinkManager.maxRefreshRateEnabledOnIPhone) {
+    return;
+  }
+  double maxFrameRate = fmax(refreshRate, 60);
+  double minFrameRate = fmax(maxFrameRate / 2, 60);
+  if (@available(iOS 15.0, *)) {
+    _displayLink.preferredFrameRateRange =
+        CAFrameRateRangeMake(forceMax ? maxFrameRate : minFrameRate, maxFrameRate, maxFrameRate);
+  } else {
+    _displayLink.preferredFramesPerSecond = maxFrameRate;
+  }
+}
+
+- (void)onDisplayLink:(CADisplayLink*)link {
+  _didSetContentsDuringThisDisplayLinkPeriod = NO;
+  // Do not pause immediately, this seems to prevent 120hz while touching.
+  if (_displayLinkPauseCountdown == 3) {
+    _displayLink.paused = YES;
```

Comment:
maybe im missing some context - i'm having trouble understanding why are we pausing display link here.

Reply:
@hellohuanlin for context please see https://github.com/flutter/engine/blob/a53ca8a17c85065cd69a6ce10d6f9bc7de9e899a/shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm .

IN short, DisplayLink ticks at a given rate and if you don't want to continually tick (because you know when you need to schedule frames) then you have to pause it.

Reply:
Displaylink on main thread is actually driving vsync. That means includes a CATransaction frame. I haven't really profiled the overhead from it, but very likely there is some, compared to main thread just sleeping on idle. The questionable part here is pausing after **3** frames. It might not be necessary - it's more of a better safe than sorry thing.